### PR TITLE
Add stats to the global redux store

### PR DESCRIPTION
**Pull request by jblz**
User Stats State
================

A module for managing the data which backs a user or site's Stats reporting features.

## Action Creators
* `createFetchActionsForModule`

### Actions
* `fetchSiteStats`

### Reducers
* `items`
* `fetchingItems`

### Selectors
* `getStatsItem`
* `isStatsItemFetching`

### Utils
* `getStatsTypesByModule`
* `getCompositeKey`
* `normalizeParams`
* `isDefined`

## Still `@TODO`:
* Persistence / injecting initial state
* Full feature parity with `StatsList`:
  * API timing
  * Error handling
  * Request retrying
* Tests
* Finish moving components away from `StatsList`

This is a work-in-progress and will be filled in as the implementation develops.

### CODE REVIEW

File: client/my-sites/stats/controller.js [hunk 1]
```diff
@@ -580,41 +584,26 @@ module.exports = {
...
+					];
 					break;
-
 				case 'videodetails':
-					summaryList = new StatsList( { statType: 'statsVideo', post: queryOptions.post, siteID: siteId, period: activeFilter.period, date: endDate, max: 0, domain: site.slug } );
-					break;
-
-				case 'searchterms':
-					summaryList = new StatsList( { siteID: siteId, statType: 'statsSearchTerms', period: activeFilter.period, date: endDate, max: 0, domain: site.slug } );
+					fetchActions = createFetchActionsForModule( module, activeFilter, endDate, { post: queryOptions.post } );
```
Comment (jblz) [inline client/my-sites/stats/controller.js]: Check to see if this `post` param applies to both actions... I'm guessing not.

TODO: don't accept the extra params to `createFetchActionsForModule` & just explicitly modify the fetch actions here for exceptions.

File: client/my-sites/stats/stats-post-detail/index.jsx [hunk 2]
```diff
@@ -67,19 +88,44 @@ export default React.createClass( {
...
 	}
 } );
+
+export default connect(
+	function mapStateToProps( state, ownProps ) {
+		const { siteID, postID, statType } = ownProps;
+		const params = {
+			siteID,
+			statType,
+			options: {
```
Comment (jblz) [inline client/my-sites/stats/stats-post-detail/index.jsx]: If we pass `postID` here as well, we can automate this in `getStatsItem` & `isStatsItemFetching`.

Ideally, we wrap `connect` in our own util function `statsConnect` (or something) that handles this boilerplate for us in each component.

File: client/state/stats/selectors.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { getCompositeKey, normalizeParams } from './utils';
+
+export function getStatsItem( state, params ) {
+	const _params = normalizeParams( params );
+	const key = getCompositeKey( _params );
```
Comment (jblz) [inline client/state/stats/selectors.js]: Note to self: try out the [deterministic-stringify](https://github.com/Automattic/wp-calypso/tree/8761dbc6ee7288a57c89e236d2b406187a6d4079/client/lib/deterministic-stringify) lib here.

**Comment by aduth** [comment]
This pull request will need to be updated with the merger of #3773:

- No need to specify custom `Makefile` in `client/state`
 - Instead, require your tests directly in `client/state/test/index.js`
- No need to specify Chai middleware (already configured in root test suite) 
- Do not call `nock.restore()` in your test tear down. Instead, call `nock.cleanAll()` (the former will disable `nock` altogether for the remainder of tests)

**Comment by lancewillett** [comment]
Needs a refresh still.

**Comment by jblz** [comment]
Closing since #5743 was merged

**Pull request closed without merging**